File: kotlin-java/src/main/kotlin/baekjoon/silver/s2/섬의개수/섬의개수.java
package baekjoon.silver.s2.섬의개수;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class 섬의개수 {

    static int[] dx = {0, 0, 1, -1, 1, 1, -1, -1};
    static int[] dy = {1, -1, 0, 0, 1, -1, 1, -1};
    static int[][] map;
    static int n = 0;
    static int m = 0;
    static boolean[][] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        while (true) {
            String[] nm = br.readLine().split(" ");
            n = Integer.parseInt(nm[1]);
            m = Integer.parseInt(nm[0]);
            int cnt = 0;

            if (n == 0 && m == 0) break;

            map = new int[n][m];
            visited = new boolean[n][m];

            for (int i = 0; i < n; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                for (int j = 0; j < m; j++) {
                    map[i][j] = Integer.parseInt(st.nextToken());
                }
            }

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (!visited[i][j] && map[i][j] == 1) {
                        cnt++;
                        dfs(i, j);
                    }
                }
            }

            sb.append(cnt).append("\n");
        }

        System.out.print(sb.toString());
    }

    static void dfs(int x, int y) {

        for (int i = 0; i < 8; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx < 0 || nx >= n || ny < 0 || ny >= m || visited[nx][ny] || map[nx][ny] == 0) continue;

            visited[nx][ny] = true;
            dfs(nx, ny);
        }
    }
}

File: Kotlin-Java/src/main/kotlin/baekjoon/silver/s2/섬의개수/섬의 개수.kt
package baekjoon.silver.s2.섬의개수

import java.util.*

val dx = arrayOf(0, 0, 1, -1, 1, -1, 1, -1)
val dy = arrayOf(1, -1, 0, 0, 1, -1, -1, 1)

var cnt = 0
var n = 0
var m = 0

lateinit var map: Array<Array<Int>>
lateinit var visited: Array<Array<Boolean>>

fun main() = with(System.`in`.bufferedReader()) {
    val sb = StringBuilder()

    while (true) {
        cnt = 0

        val nm = readLine().split(" ").map { it.toInt() }
        n = nm[1]
        m = nm[0]

        if (n == 0 && m == 0) break

        map = Array(n) { Array(m) { 0 } }
        visited = Array(n) { Array(m) { false } }

        repeat(n) { x ->
            val st = StringTokenizer(readLine())
            repeat(m) { y ->
                map[x][y] = st.nextToken().toInt()
            }
        }

        repeat(n) { x ->
            repeat(m) { y ->
                if (map[x][y] == 1 && !visited[x][y]) {
                    dfs(x, y)
                    cnt++
                }
            }
        }

        sb.append("$cnt\n")
    }

    print(sb)
}


fun dfs(x: Int, y: Int) {
    visited[x][y] = true

    for (i in 0 until 8) {
        val nx = dx[i] + x
        val ny = dy[i] + y

        if (nx < 0 || nx >= n || ny < 0 || ny >= m || visited[nx][ny] || map[nx][ny] == 0) continue

        visited[nx][ny] = true
        dfs(nx, ny)
    }
}